### PR TITLE
Bokeh 2.4 compatibility for BokehModel ipywidget

### PR DESCRIPTION
**Pull request by philippjfr**
@mattpap Could you review, merge and release?

### CODE REVIEW

File: jupyter_bokeh/widgets.py [hunk 1]
```diff
@@ -126,13 +127,13 @@ def _sync_model(self, _, content, _buffers):
             submodel = self._model.select_one({"id": content["id"]})
             descriptor = submodel.lookup(content['attr'])
             try:
-                descriptor._real_set(submodel, old, new, hint=hint, setter=self)
+                descriptor._set(submodel, old, new, hint=hint, setter=self)
```
Comment (philippjfr) [inline jupyter_bokeh/widgets.py:130]: If we could find a public API for this that would be ideal.

File: jupyter_bokeh/widgets.py [hunk 2]
```diff
@@ -126,13 +127,13 @@ def _sync_model(self, _, content, _buffers):
...
             except Exception:
                 return
             for cb in submodel._callbacks.get(attr, []):
                 cb(attr, old, new)
         elif kind == 'MessageSent':
-            self._document.apply_json_event(content["msg_data"])
+            self._document.callbacks.trigger_json_event(content["msg_data"])
```
Comment (philippjfr) [inline jupyter_bokeh/widgets.py:136]: This is frustrating, this method was public.

Comment (bryevdv) [reply jupyter_bokeh/widgets.py:136]: I thought I telegraphed these changes pretty loudly, these PRs happened over the course of several weeks. Also the stated position in the past has been "Don't worry about breaking changes, we pin Bokeh minor versions". Yes this was "public" but this code here is probably literally the only consumer of that API anywhere in the world. The real issue is that there are no downstream tests between Bokeh and `jupyter_bokeh`. I am sorry you are frustrated. I am doing the best I can.

File: jupyter_bokeh/widgets.py (hunk 1 above)
Comment (bryevdv) [reply jupyter_bokeh/widgets.py:130]: Happy consider any suggestions that come with downstream tests.

Comment (bryevdv) [reply jupyter_bokeh/widgets.py:130]: That said, I am also perfectly happy to keep it private and maintain it with tests and an understanding that `jupyter_bokeh` is part of Bokeh just maintained in a separate repo. Another overarching issue with Bokeh is that far too many things were made "public" API that are of no use  or interest to any normal users and I would very much like to reduce this surface.

File: jupyter_bokeh/widgets.py (hunk 2 above)
Comment (philippjfr) [reply jupyter_bokeh/widgets.py:136]: I'm sorry I wasn't more careful in my phrasing and I know you are very diligent with all your changes. You're right that downstream tests are missing here and also that this was telegraphed well ahead of time. I'm also trying to keep up with 10 different projects and doing the best I can 🙂

Comment (philippjfr) [reply jupyter_bokeh/widgets.py:136]: Indeed the real breakage here was in private API so this Widget would have been broken regardless.

Comment (bryevdv) [reply jupyter_bokeh/widgets.py:136]: I am going to try and prioritize CI, automation, and testing for this repo over the next few weeks.

File: jupyter_bokeh/widgets.py (hunk 1 above)
Comment (bryevdv) [reply jupyter_bokeh/widgets.py:130]: Adding: I had tried to clarify all this with the "General API" vs "Dev API" distinction, but that is just confusing for users and hard to maintain and curate, and very hard to document efectively. At this point, personally, I would like to work toward paring the public API down to only just what regular users  need to know about, and then work together on internally documenting and maintaining specific underscore APIs that are intended for cross-project use (e.g. maybe a naming convention `_internal_foo` for an internal maintained API and `_foo` for "truly private, not to rely on"

**Comment by mattpap** [comment]
@philippjfr, I published 3.0.4 with these changes.

**Pull request merged**